drain-service/drain_cp_inferencing.py

```diff
@@ -0,0 +1,269 @@
+    '''
+    global cp_predictions
+    while True:
+        cp_logs_df = await incoming_cp_logs_queue.get()
+        start_time = time.time()
+        logging.info("Received payload of size {}".format(len(cp_logs_df)))
+        logs_inferenced_results = []
+        nulog_logs = []
+        for index, row in cp_logs_df.iterrows():
+            log_message = row["masked_log"]
+            if log_message:
+                row_dict = row.to_dict()
+                result = cp_template_miner.match(log_message)
+                if result:
+                    result_template = result.get_template()
+                    prediction = cp_predictions[result_template]
+                    row_dict["anomaly_level"] = prediction
+                    row_dict["drain_control_plane_template_matched"] = result_template
+                    logs_inferenced_results.append(row_dict)
+                else:
+                    nulog_logs.append(row_dict)
+        if len(logs_inferenced_results) > 0:
+            logs_inferenced_drain_df = (pd.DataFrame(logs_inferenced_results).to_json().encode())
+            await nw.publish("anomalies", logs_inferenced_drain_df)
+        if len(nulog_logs) > 0:
+            nulog_logs_df = pd.DataFrame(nulog_logs).to_json().encode()
+            await nw.publish("nulog_cp_logs", nulog_logs_df)
+            logging.info(f"Published {len(nulog_logs)} logs to be inferenced on by Nulog.")
+        logging.info(f"{len(cp_logs_df)} logs processed in {(time.time() - start_time)} second(s)")
+
+
+async def setup_es_connection():
+    # This function will be setting up the Opensearch connection.
+    logging.info("Setting up AsyncElasticsearch")
+    return AsyncElasticsearch(
+        [ES_ENDPOINT],
+        port=9200,
+        http_auth=(ES_USERNAME, ES_PASSWORD),
+        http_compress=True,
+        max_retries=10,
+        retry_on_status={100, 400, 503},
+        retry_on_timeout=True,
+        timeout=20,
+        use_ssl=True,
+        verify_certs=False,
+        sniff_on_start=False,
+        # refresh nodes after a node fails to respond
+        sniff_on_connection_fail=True,
+        # and also every 60 seconds
+        sniffer_timeout=60,
+        sniff_timeout=10,
+    )
+
+async def update_es_logs(queue):
+    # This function will be updating Opensearch logs which were inferred on by the DRAIN model.
+    es = await setup_es_connection()
+
+    async def doc_generator_anomaly(df):
+        for index, document in df.iterrows():
+            doc_dict = document.to_dict()
+            doc_dict["doc"] = {}
+            doc_dict["doc"]["anomaly_level"] = doc_dict["anomaly_level"]
+            doc_dict["doc"]["drain_control_plane_template_matched"] = doc_dict["drain_control_plane_template_matched"]
+            del doc_dict["anomaly_level"]
+            del doc_dict["drain_control_plane_template_matched"]
+            yield doc_dict
+
+    async def doc_generator(df):
```

Comment:
can we merge these 2 functions : `doc_generator` & `doc_generator_anomaly`.  you probably just need something like:
```
if 'anomaly_level' in doc_dict:
```